Create polls

connyduck opened the pull request:
Closes #1102

- [x] allow creation of poll
- [x] honor poll limits from instanceinfo
- [x] poll support in drafts
- [x] poll support in delete and redraft

- [x] code cleanup
- [x] input validation
- [x] testing

This implementation does not support adding media and polls to the same status.

connyduck commented on app/src/main/java/com/keylesspalace/tusky/view/PollPreviewView.kt:
```diff
@@ -0,0 +1,58 @@
+
+    override fun setOnClickListener(l: OnClickListener?) {
+        super.setOnClickListener(l)
+        adapter.setOnClickListener(l)
```
This is a workaround because for some reason no clicks are registered if you set the listener on the outer layout only 🙄 (I tried a lot of other things as well)

connyduck commented:
Ok should be done now.
Only thing Im not sure about is how to show input errors? Right now the "ok" button of the dialog is blocked until the input is correct 🤔 
I tried marking the `TextInputLayout`s as error, but that takes so much space :/

Maybe a toast? Or leave it as it is?

charlag commented:
I'm afraid it doesn't save multiple "choices status"?
I also think we should disable or hide "add choice" button when the limit is reached.

Ok button sometimes doesn't do anything for me, e.g. when one of the fields is empty. I think toast or disabling button would do it.

It crashes every time for me when you try to remove an option which is focused. Probably a bug in some of the support libs?

<details>
<summary>
stacktrace
</summary>

```
2019-08-18 17:35:43.119 29172-29172/? E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.keylesspalace.tusky, PID: 29172
    java.lang.IllegalArgumentException: parameter must be a descendant of this view
        at android.view.ViewGroup.offsetRectBetweenParentAndChild(ViewGroup.java:5959)
        at android.view.ViewGroup.offsetDescendantRectToMyCoords(ViewGroup.java:5888)
        at androidx.core.widget.NestedScrollView.isWithinDeltaOfScreen(NestedScrollView.java:1383)
        at androidx.core.widget.NestedScrollView.onSizeChanged(NestedScrollView.java:1843)
        at android.view.View.sizeChange(View.java:19792)
        at android.view.View.setFrame(View.java:19753)
        at android.view.View.layout(View.java:19656)
        at android.view.ViewGroup.layout(ViewGroup.java:6075)
        at android.widget.FrameLayout.layoutChildren(FrameLayout.java:323)
        at android.widget.FrameLayout.onLayout(FrameLayout.java:261)
        at android.view.View.layout(View.java:19659)
        at android.view.ViewGroup.layout(ViewGroup.java:6075)
        at android.widget.FrameLayout.layoutChildren(FrameLayout.java:323)
        at android.widget.FrameLayout.onLayout(FrameLayout.java:261)
        at android.view.View.layout(View.java:19659)
        at android.view.ViewGroup.layout(ViewGroup.java:6075)
        at androidx.appcompat.widget.AlertDialogLayout.setChildFrame(AlertDialogLayout.java:348)
        at androidx.appcompat.widget.AlertDialogLayout.onLayout(AlertDialogLayout.java:341)
        at android.view.View.layout(View.java:19659)
        at android.view.ViewGroup.layout(ViewGroup.java:6075)
        at android.widget.FrameLayout.layoutChildren(FrameLayout.java:323)
        at android.widget.FrameLayout.onLayout(FrameLayout.java:261)
        at android.view.View.layout(View.java:19659)
        at android.view.ViewGroup.layout(ViewGroup.java:6075)
        at android.widget.FrameLayout.layoutChildren(FrameLayout.java:323)
        at android.widget.FrameLayout.onLayout(FrameLayout.java:261)
        at android.view.View.layout(View.java:19659)
        at android.view.ViewGroup.layout(ViewGroup.java:6075)
        at android.widget.FrameLayout.layoutChildren(FrameLayout.java:323)
        at android.widget.FrameLayout.onLayout(FrameLayout.java:261)
        at android.view.View.layout(View.java:19659)
        at android.view.ViewGroup.layout(ViewGroup.java:6075)
        at android.widget.LinearLayout.setChildFrame(LinearLayout.java:1791)
        at android.widget.LinearLayout.layoutVertical(LinearLayout.java:1635)
        at android.widget.LinearLayout.onLayout(LinearLayout.java:1544)
        at android.view.View.layout(View.java:19659)
        at android.view.ViewGroup.layout(ViewGroup.java:6075)
        at android.widget.FrameLayout.layoutChildren(FrameLayout.java:323)
        at android.widget.FrameLayout.onLayout(FrameLayout.java:261)
        at com.android.internal.policy.DecorView.onLayout(DecorView.java:761)
        at android.view.View.layout(View.java:19659)
        at android.view.ViewGroup.layout(ViewGroup.java:6075)
        at android.view.ViewRootImpl.performLayout(ViewRootImpl.java:2496)
        at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:2212)
        at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:1392)
        at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:6752)
        at android.view.Choreographer$CallbackRecord.run(Choreographer.java:911)
        at android.view.Choreographer.doCallbacks(Choreographer.java:723)
        at android.view.Choreographer.doFrame(Choreographer.java:658)
        at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:897)
        at android.os.Handler.handleCallback(Handler.java:790)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:164)
        at android.app.ActivityThread.main(ActivityThread.java:6494)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:440)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:807)
```

</details>

charlag commented on app/src/main/java/com/keylesspalace/tusky/ComposeActivity.java:
```diff
@@ -2033,6 +2142,12 @@ public IntentBuilder sensitive(boolean sensitive) {
         }
 
+        public IntentBuilder poll(NewPoll poll) {
+            Log.e("COMPOSE", "poll duration "+poll.getExpiresIn());
```
Log left here

connyduck commented on app/src/main/java/com/keylesspalace/tusky/view/AddPollDialog.kt:
```diff
@@ -0,0 +1,108 @@
+
+    dialog.setOnShowListener {
+        val button = dialog.getButton(AlertDialog.BUTTON_POSITIVE)
+        button.setOnClickListener {
```
lol forget to delete the old one

connyduck commented on app/src/main/java/com/keylesspalace/tusky/adapter/AddPollOptionsAdapter.kt:
```diff
@@ -0,0 +1,87 @@
+    }
+
+    fun validateInput(): Boolean {
+        if (options.contains("") || options.distinct().size != options.size) {
```
polls where options are empty are not valid

charlag replied:
Ah sorry i thought you're looking for an empty string inside string.

connyduck commented on app/src/main/java/com/keylesspalace/tusky/ComposeActivity.java:
```diff
@@ -1685,8 +1778,9 @@ private void handleCloseButton() {
         boolean contentWarningChanged = contentWarningBar.getVisibility() == View.VISIBLE &&
                 !TextUtils.isEmpty(contentWarning) && !startingContentWarning.startsWith(contentWarning.toString());
         boolean mediaChanged = !mediaQueued.isEmpty();
+        boolean pollChanged = poll != null;
```
no, here "changed" means changed from empty state

connyduck commented:
Thx for the review!

I think I addressed all issues you found & added the poll duration to the poll preview.

Yes, the code of `ComposeActivity` is a mess, I would rewrite it in Kotlin if I had the time. For now its better to not touch it so https://github.com/tuskyapp/Tusky/pull/1004 does not get too many conflicts.
